test/authenticate.user.test.js

```diff
@@ -0,0 +1,478 @@
+const
+    { closeConn, connect } = require("../config/db.server.config"),
+    { resetCollections } = require("../models/db.server.model"),
+    {
+        authenticateUser,
+        isUserAuthorized,
+        getUserAuthToken,
+        setUserAuthToken
+    } = require("../models/user.server.model"),
+    request = require('supertest'),
+    assert = require("assert"),
+    app = require('../server');
+
+/**
+ * Before each test, the app database is disconnected before the test database is connected and all data is deleted
+ */
+beforeEach(async function() {
+    const testDatabaseName = process.env.DATABASE_TEST_NAME;
+    await closeConn(); // Disconnect from the app database
+    await connect(testDatabaseName, true); // Connect to the test database
+    await resetCollections(); // reset database for testing
+});
+
+/**
+ * After all tests, all test database document data is deleted and the test database is disconnected
+ */
+after(async function() {
+    await resetCollections(); // reset database for testing
+    await closeConn(true); // Disconnect from the app database
+});
+
+/**
+ * Test successful user authentication using a matching email login and password
+ */
+describe("Authenticate user successfully with matching email and password", function() {
+    it("should return: true", function(done) {
+        const
+            email = 'test@dummy.com',
+            password = 'authentication-test';
+        request(app)
+            .post('/api/v1/users')
+            .send({
+                username: 'TestDummy',
+                displayName: 'MostValuedTest',
+                email: email,
+                password: password
+            })
+            .expect(201)
+            .end(function(err, res) {
+                if (err) done(err);
+                authenticateUser({ email: email }, password, function (result) {
+                    assert.equal(result.email === email && result.hashedPassword === password, true);
+                    done();
+                });
+            });
+    });
+});
+
+/**
+ * Test successful user authentication using a matching email login and password
+ */
+describe("Authenticate user successfully with matching username and password", function() {
+    it("should return: true", function(done) {
+        const
+            username = 'TestDummy',
+            password = 'authentication-test';
+        request(app)
+            .post('/api/v1/users')
+            .send({
+                username: username,
+                displayName: 'MostValuedTest',
+                email: 'test@dummy.com',
+                password: password
+            })
+            .expect(201)
+            .end(function(err, res) {
+                if (err) done(err);
+                authenticateUser({ username: username }, password, function (result) {
+                    assert.equal(result.username === username && result.hashedPassword === password, true);
```

Comment:
This will fail once the hashPassword method is implemented, as result.hashedPassword won't equal password.

Reply:
The feature isn't yet implemented and I would assume whoever is implementing the password hash feature would apply necessary changes here